File: data-structures/non-linear/binary-tree/heap/min-heap/MinHeap.java
import java.util.ArrayList;

class MinHeap {
    ArrayList<Integer> heap;

    public MinHeap() {
        heap = new ArrayList<>();
    }

    int parent(int i) {
        return (i - 1) / 2;
    }

    int leftChild(int i) {
        return 2 * i + 1;
    }

    int rightChild(int i) {
        return 2 * i + 2;
    }

    void upHeap(int i) {
        while (i > 0 && heap.get(i) < heap.get(parent(i))) {
            swap(i, parent(i));
            i = parent(i);
        }
    }

    void downHeap(int i) {
        int smallest = i;
        int l = leftChild(i);
        int r = rightChild(i);

        if (l < heap.size() && heap.get(l) < heap.get(smallest)) {
            smallest = l;
        }

        if (r < heap.size() && heap.get(r) < heap.get(smallest)) {
            smallest = r;
        }

        if (smallest != i) {
            swap(i, smallest);
            downHeap(smallest);
        }
    }

    void insert(int key) {
        heap.add(key);
        upHeap(heap.size() - 1);
    }

    int getMin() {
        if (empty()) {
            System.out.println("Heap is empty!");
            return -1;
        }
        return heap.get(0);
    }

    int extractMin() {
        if (empty()) {
            System.out.println("Heap is empty!");
            return -1;
        }

        int root = heap.get(0);

        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);

        if (!empty()) {
            downHeap(0);
        }

        return root;
    }

    void buildHeap(ArrayList<Integer> arr) {
        heap = new ArrayList<>(arr);
        for (int i = parent(heap.size() - 1); i >= 0; i--) {
            downHeap(i);
        }
    }

    void printHeap() {
        for (int x : heap) {
            System.out.print(x + " ");
        }
        System.out.println();
    }

    boolean empty() {
        return heap.isEmpty();
    }

    int size() {
        return heap.size();
    }

    void swap(int i, int j) {
        int temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }
}

File: data-structures/non-linear/binary-tree/heap/min-heap/MinHeap.ts
class MinHeap {
	private heap: Array<number> = [];

	private getParent(index: number): number {
		return (index - 1) / 2;
	}

	private getLeftChild(index: number): number {
		return 2 * index + 1;
	}

	private getRightChild(index: number): number {
		return 2 * index + 2;
	}

	private upHeap(i: number): void {
		let temp: number;
		while (i > 0 && this.heap[i] < this.heap[this.getParent(i)]) {
			temp = this.heap[i];
			this.heap[i] = this.heap[this.getParent(i)];
			this.heap[this.getParent(i)] = temp;
			i = this.getParent(i);
		}
	}

	private downHeap(i: number): void {
		let smallest: number = i;
		const rightChild: number = this.getRightChild(i);
		const leftChild: number = this.getLeftChild(i);

		if (
			leftChild < this.heap.length &&
			this.heap[leftChild] < this.heap[smallest]
		) {
			smallest = leftChild;
		}

		if (
			rightChild < this.heap.length &&
			this.heap[rightChild] < this.heap[smallest]
		) {
			smallest = rightChild;
		}

		if (smallest !== i) {
			const temp: number = this.heap[smallest];
			this.heap[smallest] = this.heap[i];
			this.heap[i] = temp;
			this.downHeap(smallest);
		}
	}

	public insert(key: number): void {
		this.heap.push(key);
		this.upHeap(this.heap.length - 1);
	}

	public getMin(): number {
		if (this.isEmpty()) return -1;

		return this.heap[0];
	}

	public extractMin(): number {
		if (this.isEmpty()) return -1;

		const min: number = this.heap[0];

		//@ts-expect-error If the heap is empty the function terminates early, array.pop returns undefined when empty array
		this.heap[0] = this.heap.pop();

		//check if this.empty is required
		if (!this.isEmpty()) {
			this.downHeap(0);
		}

		return min;
	}

	public buildHeap(heapArray: Array<number>): void {
		this.heap = heapArray;

		for (let i = this.getParent(this.heap.length - 1); i >= 0; i--) {
			this.downHeap(i);
		}
	}

	public printHeap(): void {
		for (const x of this.heap) {
			console.log(`${x}`);
		}
	}

	public isEmpty(): boolean {
		return this.heap.length === 0;
	}

	public size(): number {
		return this.heap.length;
	}
}

export { MinHeap };
